staging/src/k8s.io/apimachinery/pkg/runtime/converter.go

```diff
@@ -358,38 +497,57 @@ func pointerFromUnstructured(sv, dv reflect.Value) error {
 	dv.Set(reflect.New(dt.Elem()))
 	switch st.Kind() {
 	case reflect.Ptr, reflect.Interface:
-		return fromUnstructured(sv.Elem(), dv.Elem())
+		return fromUnstructured(sv.Elem(), dv.Elem(), ctx)
 	default:
-		return fromUnstructured(sv, dv.Elem())
+		return fromUnstructured(sv, dv.Elem(), ctx)
 	}
 }
 
-func structFromUnstructured(sv, dv reflect.Value) error {
+func structFromUnstructured(sv, dv reflect.Value, ctx *fromUnstructuredContext) error {
 	st, dt := sv.Type(), dv.Type()
 	if st.Kind() != reflect.Map {
 		return fmt.Errorf("cannot restore struct from: %v", st.Kind())
 	}
 
+	pathLen := len(ctx.parentPath)
+	defer func() {
+		ctx.parentPath = ctx.parentPath[:pathLen]
+	}()
+	ctx.pushMatchedKeyTracker()
 	for i := 0; i < dt.NumField(); i++ {
 		fieldInfo := fieldInfoFromField(dt, i)
 		fv := dv.Field(i)
 
+		origInlined := ctx.isInlined
```

Comment:
move this outside the for loop and defer restoring the original so we can be sure we restore it even if we return early for an error

```go
pathLen := len(ctx.parentPath)
origInlined := ctx.isInlined
defer func() {
	ctx.parentPath = ctx.parentPath[:pathLen]
	ctx.isInlined = origInlined
}()
for i := 0; i < dt.NumField(); i++ {
...
```